File: zmq_testing/zmq_thread_test.py
import threading
import zmq

context = zmq.Context()
lock = threading.Semaphore()
do_continue = True


def eater():
    global do_continue
    socket = context.socket(zmq.PAIR)
    socket.bind("inproc://a")
    poller = zmq.Poller()
    poller.register(socket, zmq.POLLIN)
    lock.release()
    while do_continue:
        print "Going to poll"
        try:
           socks = dict(poller.poll())
        except zmq.ContextTerminated, e:
            return;
        except Exception, e:
            print e


        if socket in socks:
            message = socket.recv_string()
            print "Received: ", message
            if message != 'add' and message != '':
                print "ERROR: Received bad message: ", message
            socket.send_string("OK")
    print "yellow"
    socket.close()

def feeder():
    socket = context.socket(zmq.PAIR)
    socket.connect("inproc://a")
    for i in range(0,10):
        sender_string = "add"
        print i%2
        if i %2 == 0:
            sender_string = ""
        print "Sending string of size: ", len(sender_string), " for iteration:", i
        socket.send_string(sender_string)

        response = socket.recv_string()
        print "Got response: ", response
    socket.close()


def main():
    global do_continue
    lock.acquire()
    thread1 = threading.Thread(target=eater)
    thread1.start()
    lock.acquire()
    thread2 = threading.Thread(target=feeder)
    thread2.start()
    thread2.join()
    print "------------------------------between 2"
    thread3 = threading.Thread(target=feeder)
    thread3.start()
    thread3.join()
    do_continue = False
    context.term()

if __name__ == "__main__":
    main()

File: zmq_testing/zmq_thread_test.cc
#include <sys/types.h>
#include <time.h>
#include <zmq.hpp>



#include <map>
#include <thread>
#include <mutex>
#include <iostream>


std::mutex wait_lock;
bool do_continue = true;
zmq::context_t inproc_context(1);


void eat_thread() {
  zmq::socket_t inproc_socket(inproc_context, ZMQ_PAIR);
  zmq::pollitem_t poll_items[512];

  try {
    inproc_socket.bind("inproc://a");
  } catch(zmq::error_t e) {
    std::cout << "Error on inproc bindq : " << e.what() << "\n";
    return;
  }
  int major, minor, patch;
  zmq::version (&major, &minor, &patch);
  std::cout << "Current 0MQ version is " << major << "." << minor << "."
            << patch << "\n";

  poll_items[0] = {inproc_socket, 0, ZMQ_POLLIN, 0};
  wait_lock.unlock();

  while (do_continue) {
    uint32_t current_count = 1;
    std::cout << "we are going to test out our subscriptions!"
              << " : " << current_count << "\n";
    int rc;
    try {
      if(!(rc=zmq::poll(&poll_items[0], current_count,-1)))  {
        continue;
      }
    }
    catch(zmq::error_t& e) {
      //XXX what to do here? need to be smarter
      if (e.num() == ETERM) {
        return;
      }
      std::cout << "Error received in zmq_poll: " << e.what()
                 << " (" << e.num() << ")\n";
      continue;
    }
    if (poll_items[0].revents & ZMQ_POLLIN) {
      zmq::message_t type;
      inproc_socket.recv(&type);

      std::string cmd_type(static_cast<char*>(type.data()), type.size());
      std::cout << "Received: " << type.size() << "\n";

      if (cmd_type != "add" && cmd_type != "") {
        std::cout <<  "recieved: '" << static_cast<char*>(type.data())
                  << " AKA: '" << cmd_type << "'\n";
      }
      zmq::message_t resp(2);
      memcpy(resp.data(), "OK", 2);
      inproc_socket.send(resp);
    }
  }
}


void feeder() {
  zmq::socket_t inproc_socket(inproc_context, ZMQ_PAIR);
  try {
    inproc_socket.connect("inproc://a");
  } catch(zmq::error_t e) {
    std::cout << "Error on zmq inproc connect: " << e.what() << "\n";
    return;
  }
  std::string command("add");
  std::string publisher_host("localhost:9988");
  for(int i = 0; i < 10; i++) {
    if (i%2 == 0) {
      std::string filter_name("");
      std::cout << filter_name.size() << "\n";
      zmq::message_t cmd(filter_name.size());
      memcpy(cmd.data(), filter_name.data(), filter_name.size());
      inproc_socket.send(cmd);
    } else {
      std::string filter_name("add");
      std::cout << filter_name.size() << "\n";
      zmq::message_t cmd(filter_name.size());
      memcpy(cmd.data(), filter_name.data(), filter_name.size());
      inproc_socket.send(cmd);
    }
    zmq::message_t resp;
    inproc_socket.recv(&resp);

    std::string response(static_cast<char*>(resp.data()), resp.size());
    std::cout << "Got response: " << response << "\n";
  }
}

int main(int argc, char** argv)
{
  wait_lock.lock();
  std::thread t(eat_thread);
  t.detach();
  wait_lock.lock();
  std::thread t1(feeder);
  t1.join();
  std::cout << "between two threads\n";
  std::thread t2(feeder);
  t2.join();
  do_continue = false;
  sleep(1);
}




